ksqldb-engine/src/main/java/io/confluent/ksql/planner/LogicalPlanner.java

```diff
@@ -460,49 +460,81 @@ private FlatMapNode buildFlatMapNode(final PlanNode sourcePlanNode) {
     return new FlatMapNode(new PlanNodeId("FlatMap"), sourcePlanNode, metaStore, analysis);
   }
 
-  private PlanNode buildSourceForJoin(
-      final AliasedDataSource source,
+  private PlanNode prepareSourceForJoin(
+      final JoinTree.Node node,
+      final PlanNode joinSource,
       final String side,
       final Expression joinExpression,
-      final boolean isWindowed
+      final boolean isForeignKeyJoin
   ) {
-    final DataSourceNode sourceNode = new DataSourceNode(
-        new PlanNodeId("KafkaTopic_" + side),
-        source.getDataSource(),
-        source.getAlias(),
-        isWindowed
-    );
+    if (node instanceof JoinTree.Join) {
+      return prepareSourceForJoin(
+          (JoinTree.Join) node,
+          joinSource,
+          side,
+          joinExpression,
+          isForeignKeyJoin
+      );
+    } else {
+      return prepareSourceForJoin(
+          (DataSourceNode) joinSource,
+          side,
+          joinExpression,
+          isForeignKeyJoin
+      );
+    }
+  }
 
-    // it is always safe to build the repartition node - this operation will be
-    // a no-op if a repartition is not required. if the source is a table, and
-    // a repartition is needed, then an exception will be thrown
-    final VisitParentExpressionVisitor<Optional<Expression>, Context<Void>> rewriter =
-        new VisitParentExpressionVisitor<Optional<Expression>, Context<Void>>(Optional.empty()) {
-          @Override
-          public Optional<Expression> visitQualifiedColumnReference(
-              final QualifiedColumnReferenceExp node,
-              final Context<Void> ctx
-          ) {
-            return Optional.of(new UnqualifiedColumnReferenceExp(node.getColumnName()));
-          }
-        };
-
-    final PlanNode repartition =
-        buildInternalRepartitionNode(sourceNode, side, joinExpression, rewriter::process);
+  private PlanNode prepareSourceForJoin(
+      final DataSourceNode sourceNode,
+      final String side,
+      final Expression joinExpression,
+      final boolean isForeignKeyJoin
+  ) {
+    final PlanNode preProjectNode;
+    if (isForeignKeyJoin) {
+      // we do not need to repartition for foreign key joins, as FK joins do not
+      // have co-partitioning requirements
+      preProjectNode = sourceNode;
+    } else {
+      // it is always safe to build the repartition node - this operation will be
+      // a no-op if a repartition is not required. if the source is a table, and
+      // a repartition is needed, then an exception will be thrown
```

Comment:
ksqlDB does not allow repartitioning tables as a safeguard to prevent users from unintentionally changing their table semantics, since ordering of records across partitions is not preserved. Users can work around this limitation by declaring a stream over the topic and repartitioning the stream if they wish.

Reply:
I would not say it's about "ordering" but about fundamental semantics: `PARTITION BY` changes the primary key in ksqlDB (it has slightly different semantics than in an RDBMS), similar to `GROUP BY` and for `GROUP BY` for all non-grouping columns an aggregation function is applied to "merge" values (of a column across all rows of a group) into a single value. For `PARTITION BY` there is no such mechanism: multiple rows could map to the same "new primary key" and thus the system cannot compute a "target row" because it's not well defined.

Of course we could pick a "random" one and then the ordering statement from Victoria applies: if we pick a random one, it would be non-deterministic and we should not allow non-deterministic operations. (That is also my concern with regard to `LATEST_BY_KEY` that allows users to repartition the table using a `GROUP BY` in a basically deterministic manner.)